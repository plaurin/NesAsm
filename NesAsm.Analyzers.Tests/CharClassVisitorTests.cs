using ApprovalTests;
using ApprovalTests.Reporters;
using NesAsm.Analyzers.Visitors;

namespace NesAsm.Analyzers.Tests;

[UseReporter(typeof(VisualStudioReporter))]
public class CharClassVisitorTests
{
    [Fact]
    public void Test()
    {
        var result = CharClassVisitor.Process(@"C:\Users\pasca\Dev\GitHub\NesAsm\NesAsm.Example\game1.png");

        Approvals.Verify(result);
    }

    [Fact]
    public void TestImportCharScript()
    {
        TestScript("ImportCharDefinition.cs");
    }

    private static void TestScript(params string[] filenames)
    {
        var generator = new CharGenerator();

        var sourcePaths = filenames.Select(filename => $@"..\..\..\TestFiles\{filename}").ToArray();

        GeneratorTestUtilities.TestGenerator(generator, sourcePaths);
    }
}