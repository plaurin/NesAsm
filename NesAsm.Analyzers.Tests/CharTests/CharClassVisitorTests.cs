using ApprovalTests.Reporters;

namespace NesAsm.Analyzers.Tests.CharTests;

[UseReporter(typeof(VisualStudioReporter))]
public class CharClassVisitorTests
{
    [Fact]
    public void TestImportCharScript()
    {
        TestScript("ImportCharDefinition.cs");
    }

    [Fact]
    public void TestInvalidSizeCharScript()
    {
        TestScript("InvalidSize.cs");
    }

    [Fact]
    public void TestMoreThanFourPalettesCharScript()
    {
        TestScript("MoreThanFourPalettes.cs");
    }

    [Fact]
    public void TestMoreThanThreeNonDefaultColorCharScript()
    {
        TestScript("MoreThanThreeNonDefaultColor.cs");
    }

    [Fact]
    public void TestColorMismatchCharScript()
    {
        TestScript("ColorMismatch.cs");
    }

    private static void TestScript(params string[] filenames)
    {
        var generator = new CharGenerator();

        var sourcePaths = filenames.Select(filename => $@"..\..\..\CharTests\TestFiles\{filename}").ToArray();

        GeneratorTestUtilities.TestGenerator(generator, sourcePaths);
    }
}