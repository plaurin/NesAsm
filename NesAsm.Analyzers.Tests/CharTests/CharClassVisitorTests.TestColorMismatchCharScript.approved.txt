- Main diagnostics:
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 59, 2, 164 is not a valid NES color (using Mesen color palette), use the closest color RGB 59, 0, 164 instead (position x: 13, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 110, 2, 64 is not a valid NES color (using Mesen color palette), use the closest color RGB 110, 0, 64 instead (position x: 15, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 44, 136 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 42, 136 instead (position x: 12, y: 4)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 232, 202, 255 is not a valid NES color (using Mesen color palette), use the closest color RGB 232, 200, 255 instead (position x: 21, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 192, 225, 255 is not a valid NES color (using Mesen color palette), use the closest color RGB 192, 223, 255 instead (position x: 19, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 254, 206, 197 is not a valid NES color (using Mesen color palette), use the closest color RGB 254, 204, 197 instead (position x: 25, y: 4)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 126, 141 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 124, 141 instead (position x: 31, y: 3)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 84, 0 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 82, 0 instead (position x: 33, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 51, 55, 0 is not a valid NES color (using Mesen color palette), use the closest color RGB 51, 53, 0 instead (position x: 31, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 79, 81, 79 is not a valid NES color (using Mesen color palette), use the closest color RGB 79, 79, 79 instead (position x: 41, y: 1)
- SyntaxTrees Count: 2
- Output diagnostics:
- RunResult.GeneratedTrees Count: 1
- RunResults diagnostics:
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 59, 2, 164 is not a valid NES color (using Mesen color palette), use the closest color RGB 59, 0, 164 instead (position x: 13, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 110, 2, 64 is not a valid NES color (using Mesen color palette), use the closest color RGB 110, 0, 64 instead (position x: 15, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 44, 136 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 42, 136 instead (position x: 12, y: 4)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 232, 202, 255 is not a valid NES color (using Mesen color palette), use the closest color RGB 232, 200, 255 instead (position x: 21, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 192, 225, 255 is not a valid NES color (using Mesen color palette), use the closest color RGB 192, 223, 255 instead (position x: 19, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 254, 206, 197 is not a valid NES color (using Mesen color palette), use the closest color RGB 254, 204, 197 instead (position x: 25, y: 4)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 126, 141 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 124, 141 instead (position x: 31, y: 3)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 84, 0 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 82, 0 instead (position x: 33, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 51, 55, 0 is not a valid NES color (using Mesen color palette), use the closest color RGB 51, 53, 0 instead (position x: 31, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 79, 81, 79 is not a valid NES color (using Mesen color palette), use the closest color RGB 79, 79, 79 instead (position x: 41, y: 1)
- RunResult[0].Generator: Microsoft.CodeAnalysis.IncrementalGeneratorWrapper
- RunResult[0] diagnostics:
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 59, 2, 164 is not a valid NES color (using Mesen color palette), use the closest color RGB 59, 0, 164 instead (position x: 13, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 110, 2, 64 is not a valid NES color (using Mesen color palette), use the closest color RGB 110, 0, 64 instead (position x: 15, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 44, 136 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 42, 136 instead (position x: 12, y: 4)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 232, 202, 255 is not a valid NES color (using Mesen color palette), use the closest color RGB 232, 200, 255 instead (position x: 21, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 192, 225, 255 is not a valid NES color (using Mesen color palette), use the closest color RGB 192, 223, 255 instead (position x: 19, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 254, 206, 197 is not a valid NES color (using Mesen color palette), use the closest color RGB 254, 204, 197 instead (position x: 25, y: 4)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 126, 141 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 124, 141 instead (position x: 31, y: 3)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 0, 84, 0 is not a valid NES color (using Mesen color palette), use the closest color RGB 0, 82, 0 instead (position x: 33, y: 5)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 51, 55, 0 is not a valid NES color (using Mesen color palette), use the closest color RGB 51, 53, 0 instead (position x: 31, y: 7)
..\..\..\CharTests\TestFiles\ColorMismatch.cs(6,2): warning NA1004: Color RGB 79, 81, 79 is not a valid NES color (using Mesen color palette), use the closest color RGB 79, 79, 79 instead (position x: 41, y: 1)
- RunResult[0].GeneratedSources Count: 1
- RunResult[0].Exception: 
- RunResult[0].GeneratedSources[0].HintName: Asm.ColorMismatch.cs
- RunResult[0].GeneratedSources[0].SourceText:
/* Generator Asm code in file Asm.ColorMismatch.s
; Auto generated code using the NesAsm project
.scope ColorMismatch

.segment "CHARS"

  ; Sprite Tile 0
  .byte 0, 0, 0, 0, 0, 0, 0, 0
  .byte 0, 0, 0, 0, 0, 0, 0, 0

  ; Sprite Tile 1
  .byte %00001000
  .byte %00001100
  .byte %00001100
  .byte %11111100
  .byte %11111100
  .byte %00001100
  .byte %00001100
  .byte %00001000

  .byte %00000000
  .byte %00000000
  .byte %00000010
  .byte %11110011
  .byte %11110011
  .byte %00000010
  .byte %00000000
  .byte %00000000


  ; Sprite Tile 2
  .byte %00010000
  .byte %00110000
  .byte %00110000
  .byte %00111111
  .byte %00111111
  .byte %00110000
  .byte %00110000
  .byte %00010000

  .byte %00000000
  .byte %00000000
  .byte %01000000
  .byte %11001111
  .byte %11001111
  .byte %01000000
  .byte %00000000
  .byte %00000000


  ; Sprite Tile 3
  .byte %00011000
  .byte %00011000
  .byte %00011000
  .byte %00011000
  .byte %00000000
  .byte %00000000
  .byte %00111100
  .byte %00011000

  .byte %00000000
  .byte %00000000
  .byte %00000000
  .byte %00000000
  .byte %11111111
  .byte %01111110
  .byte %00111100
  .byte %00011000


  ; Sprite Tile 4
  .byte %00011000
  .byte %00111100
  .byte %00000000
  .byte %00000000
  .byte %00011000
  .byte %00011000
  .byte %00011000
  .byte %00011000

  .byte %00000000
  .byte %00000000
  .byte %01111110
  .byte %11111111
  .byte %00011000
  .byte %00011000
  .byte %00011000
  .byte %00011000


.segment "CODE"

sprite_palettes:
  ; Sprite Palette 0
  .byte 0F, 03, 05, 01

  ; Sprite Palette 1
  .byte 0F, 33, 31, 36

  ; Sprite Palette 2
  .byte 0F, 1C, 0A, 08

  ; Sprite Palette 3
  .byte 0F, 2D, 3A, 28

.endscope


*/
